### PR TITLE
FIX: query() and isin() fail on period type column

### LINKED ISSUE #54199: BUG: `.query()` and `isin()` fail on period type column
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame(
    dict(
        Date=pd.period_range("2000-01-01", periods=10, name="Date"),
        Value=range(10),
    ),
)

result1 = df[df.Date == "2000-01-06"]  # Correct
result2 = df.query("Date == '2000-01-06'")  # Empty!

# To show that the period type is the problem
df.Date = df.Date.dt.to_timestamp()

result3 = df[df.Date == "2000-01-06"]  # Correct
result4 = df.query("Date == '2000-01-06'")  # Correct
```

### Issue Description

I would expect `query("Date == '2000-01-06'")` to return the same regardless of whether the `Date` column was `datetime` or `period`, given that `df.Date == "2000-01-06"` returns the same in both cases.

Also, `>=`, `<`, etc. work as expected for the period column, it's just `==` that is wrong.

I dug a little deeper and it seems that `query` in this case is using `.isin` to evaluate `==`, so the issue also shows in this code without any `eval` going on.
```py
pd.period_range("2000-01-01", periods=3).isin(['2000-01-02'])  # False, False, False
pd.date_range("2000-01-01", periods=3).isin(['2000-01-02'])    # False,  True, False

pd.period_range("2000-01-01", periods=3) == '2000-01-02'       # False,  True, False
```

From what I can tell, `isin` will attempt to turn the passed list into a `PeriodArray` here:
https://github.com/pandas-dev/pandas/blob/9be48ef953fe080c68440b21cfd6d232d1537ba7/pandas/core/arrays/datetimelike.py#L766

But it will fail because it can't work out the freq, so reverts to just doing element-wise equality which compares periods and strings and so everything is `False`.

Maybe a check here to see if this is `PeriodArray` and if so pass `self.freq`, or is there something like `period_array_like` that will copy the freq/dtype from one period array to a new one.

Is it worth questioning why `.query()` uses `isin` to handle `==` in the first place? I have no idea, I'm sure there's a good reason.

### Expected Behavior

Clear from the above?

### Installed Versions

`pd.show_versions()` is still buggy, doesn't work on my machine.

I'm on version `2.0.3`.

### PR DESCRIPTION
**Pull request by dannyi96**
- [X] closes #54199
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -501,6 +502,8 @@ def isin(comps: ListLike, values: ListLike) -> npt.NDArray[np.bool_]:
     comps_array = extract_array(comps_array, extract_numpy=True)
     if not isinstance(comps_array, np.ndarray):
         # i.e. Extension Array
+        if isinstance(comps_array.dtype, PeriodDtype):
+            comps_array = comps_array.to_timestamp()
         return comps_array.isin(values)
```
Comment (dannyi96) [inline pandas/core/algorithms.py:508]: 2 testcases fail post this change, looking into it
```
FAILED pandas/tests/indexes/test_datetimelike.py::TestDatetimeLike::test_isin[simple_index0] - TypeError: Passing PeriodDtype data is invalid. Use `data.to_timestamp()` instead
FAILED pandas/tests/series/methods/test_isin.py::TestSeriesIsIn::test_isin_period_freq_mismatch - TypeError: Passing PeriodDtype data is invalid. Use `data.to_timestamp()` instead
```

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by dannyi96** [comment]
hi @mroeschke , sure.
Life got in the way. Will look into this in coming days

**Comment by dannyi96** [comment]
hi @mroeschke , I've made some changes to address the failing testcases and pushed to this branch `dannyi96:period_type_issue`.
Do we reopen this PR (or) create a new one and discard this ?

**Pull request closed without merging**